processing/src/main/java/io/druid/query/groupby/epinephelinae/BufferGrouper.java

```diff
@@ -0,0 +1,457 @@
+      }
+
+      @Override
+      public Integer set(int index, Integer element)
+      {
+        final Integer oldValue = get(index);
+        buffer.putInt(tableArenaSize + index * Ints.BYTES, element);
+        return oldValue;
+      }
+
+      @Override
+      public int size()
+      {
+        return size;
+      }
+    };
+
+    if (sorted) {
+      final KeyComparator comparator = keySerde.comparator();
+
+      // Sort offsets in-place.
+      Collections.sort(
+          wrappedOffsets,
+          new Comparator<Integer>()
+          {
+            @Override
+            public int compare(Integer lhs, Integer rhs)
+            {
+              return comparator.compare(
+                  tableBuffer,
+                  tableBuffer,
+                  lhs + HASH_SIZE,
+                  rhs + HASH_SIZE
+              );
+            }
+          }
+      );
+
+      return new Iterator<Entry<KeyType>>()
+      {
+        int curr = 0;
+
+        @Override
+        public boolean hasNext()
+        {
+          return curr < size;
+        }
+
+        @Override
+        public Entry<KeyType> next()
+        {
+          return bucketEntryForOffset(wrappedOffsets.get(curr++));
+        }
+
+        @Override
+        public void remove()
+        {
+          throw new UnsupportedOperationException();
+        }
+      };
+    } else {
+      // Unsorted iterator
+      return new Iterator<Entry<KeyType>>()
+      {
+        int curr = 0;
+
+        @Override
+        public boolean hasNext()
+        {
+          return curr < size;
+        }
+
+        @Override
+        public Entry<KeyType> next()
+        {
+          final int offset = buffer.getInt(tableArenaSize + curr * Ints.BYTES);
+          final Entry<KeyType> entry = bucketEntryForOffset(offset);
+          curr++;
+
+          return entry;
+        }
+
+        @Override
+        public void remove()
+        {
+          throw new UnsupportedOperationException();
+        }
+      };
+    }
+  }
+
+  @Override
+  public void close()
+  {
+    for (BufferAggregator aggregator : aggregators) {
+      aggregator.close();
+    }
+  }
+
+  private boolean isUsed(final int bucket)
+  {
+    return (tableBuffer.get(bucket * bucketSize) & 0x80) == 0x80;
+  }
+
+  private Entry<KeyType> bucketEntryForOffset(final int bucketOffset)
+  {
+    final KeyType key = keySerde.fromByteBuffer(tableBuffer, bucketOffset + HASH_SIZE);
+    final Object[] values = new Object[aggregators.length];
+    for (int i = 0; i < aggregators.length; i++) {
+      values[i] = aggregators[i].get(tableBuffer, bucketOffset + aggregatorOffsets[i]);
+    }
+
+    return new Entry<>(key, values);
+  }
+
+  private void growIfPossible()
+  {
+    final int newBuckets = buckets * 2;
+    final int newMaxSize = maxSizeForBuckets(newBuckets);
+
+    if (newBuckets > (tableArenaSize - tableStart) / bucketSize) {
+      // Not enough space left to grow
+      return;
```

Comment:
Probably not, this is expected to happen during queries from time to time and is handled elsewhere (will cause early emission during single-segment processing or a disk spill during merging). It would just cause noise in the logs.